### PR TITLE
[#545] Extend the transaction generators to take fees into account

### LINKED ISSUE #545: Extend the transaction generators to take fees into account
We need to take fees into account in the generators. This means that the generators should take into account the minimum fee, determined via the protocol parameters `a` and `b`. The transaction generators must produce transactions that can cover the minimum fee. This in turn requires that the UTxO has a balance high enough to allow the production of transactions that can cover the minimum fee.

### PR DESCRIPTION
**Pull request by ruhatch**
Closes #545

### CODE REVIEW

File: byron/ledger/executable-spec/src/Ledger/UTxO.hs [hunk 1]
```diff
@@ -68,11 +71,11 @@ fromTxOuts = UTxO . Map.fromList . fmap (\out -> (TxIn (mkId out) 0, out))
 data Tx = Tx
   { inputs  :: [TxIn]
   , outputs :: [TxOut]
-  } deriving (Eq, Show, Ord, Generic, Hashable)
+  } deriving (Eq, Show, Ord, Generic, Hashable, HasTypeReps)
```
Comment (dnadales) [inline byron/ledger/executable-spec/src/Ledger/UTxO.hs]: Yeah! :pray:

File: byron/ledger/executable-spec/src/Ledger/UTxO.hs [hunk 2]
```diff
@@ -68,11 +71,11 @@ fromTxOuts = UTxO . Map.fromList . fmap (\out -> (TxIn (mkId out) 0, out))
...
 
-instance HasTypeReps Tx where
-  typeReps x@(Tx inputs outputs)
-    = typeOf x <| typeOf inputs <| typeOf outputs <| empty
+-- instance HasTypeReps Tx where
```
Comment (dnadales) [inline byron/ledger/executable-spec/src/Ledger/UTxO.hs]: Kill with :fire: ?

File: byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs [hunk 3]
```diff
@@ -191,20 +200,22 @@ shrinkLeftPreserving inValue outValue modifyOutValue (xs1, ys1) = do
...
-    go 0  x        = x
-    go _  []       = []
-    go n' (x : xs) = if getVal x > n'
-      then modifyVal (subtract n') x : xs
-      else go (n' - getVal x) xs
+
+-- | Remove total value from a list, removing from the front
+subFromList
+  :: (Num n, Ord n)
+  => n
```
Comment (dnadales) [inline byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs]: You know what I'm going to say: can you add some comments on the function parameters? :pray:

File: byron/ledger/formal-spec/utxo.tex [hunk 4]
```diff
@@ -82,9 +82,9 @@ \section{UTxO}
       \fun{txbody} & \Tx \to \powerset{\TxIn} \times (\Ix \mapsto \TxOut)
                                   & \text{transaction body}\\
       %
-      \fun{a} & \PPMMap \to \mathbb{Z} & \text{minumum fee factor}\\
+      \fun{a} & \PPMMap \to \mathbb{Z} & \text{minimum fee constant}\\
```
Comment (dnadales) [inline byron/ledger/formal-spec/utxo.tex]: :pray:

File: byron/ledger/executable-spec/src/Ledger/UTxO.hs (hunk 2 above)
Comment (ruhatch) [reply byron/ledger/executable-spec/src/Ledger/UTxO.hs]: I thought I'd keep it for writing the property tests and showing this version didn't work, but maybe that's a pointless test and we should just test that the new impl works!

Comment (dnadales) [reply byron/ledger/executable-spec/src/Ledger/UTxO.hs]: Oh, also fine with me :)

**Pull request merged**